### ISSUE TITLE
Assigning lambda to kwarg-only Protocol doesn't pass strict typechecking

### ISSUE THREAD

**Issue by brandon-leapyear** [microsoft/pyright#3347]
pyright should use the information in the Protocol to infer the lambda type:
```python
# pyright: strict

from typing import Protocol

class MyProtocol(Protocol):
    def __call__(self, *, foo: int) -> str:
        ...

x: MyProtocol = lambda foo: str(foo)
```
this currently fails with
```
  foo.py:9:33 - error: Argument type is unknown
    Argument corresponds to parameter "object" in function "__new__" (reportUnknownArgumentType)
  foo.py:9:1 - error: Type of "x" is partially unknown
    Type of "x" is "(foo: Unknown) -> str" (reportUnknownVariableType)
  foo.py:9:24 - error: Type of parameter "foo" is unknown (reportUnknownLambdaType)
```
It would work if I made the lambda kwarg-only `lambda *, foo`, but that doesn't seem necessary for me; a function that can be called with/without kwargs should be assignable to a function that can be called with only kwargs.

This is a contrived example for our actual issue, where we'd like to return a constant, regardless of the args passed to the protocol:
```python
class MyProtocol(Protocol):
    def __call__(self, *, a: ..., b: ..., c: ..., ...) -> str: ...

x: MyProtocol = lambda **kwargs: 'hardcoded'
```
We do this in testing, but this gives the same errors about `kwargs` being unknown

**Comment by erictraut** [comment]
The bidirectional type inference logic for lambdas does not currently handle this case. The logic is already complex, and I'm reluctant to add more complexity to handle what I think is an extreme edge case. It looks like you've already identified a reasonable workaround. 

I'm going to close this, but we might reconsider in the future if we receive signal that it's a case that other pyright users hit.

**Comment by brandon-leapyear** [comment]
To be clear, the workaround doesnt work for the **kwargs case. My current
workaround is to just convert the lambda into a helper function, which is
fine for now, although less ideal.
